### ISSUE TITLE
Is it safe to use different boost::network::uri::uri instances in different threads?

### ISSUE THREAD

**Issue by tapapax** [cpp-netlib/cpp-netlib#571]
I think it is not, because of this line (https://github.com/cpp-netlib/uri/blob/143d4264b8db77cf5b77f8975b3ec8c118bb8a80/src/detail/uri_parse.cpp#L250):
```c++
bool parse(uri::string_type::iterator first,
               uri::string_type::iterator last,
               uri_parts<uri::string_type::iterator> &parts) {
      namespace qi = boost::spirit::qi;
      static uri_grammar<uri::string_type> grammar;      <=========== THIS ONE
      bool is_valid = qi::parse(first, last, grammar, parts);
      return is_valid && (first == last);
}
```

This function is used from uri constructors without any syncronization. You can easily reproduce this issue with this code (it is crashing in msvc2013 - v120):
```c++
void ThreadFunc(const std::string& url) {
  boost::network::uri::uri t(url);
}

int WINAPI _tWinMain(HINSTANCE hInstance, HINSTANCE /*hPrevInstance*/,
                     LPTSTR lpstrCmdLine,
                     int nCmdShow) {
  boost::thread thread1(ThreadFunc, "http://abcd.com/");
  boost::thread thread2(ThreadFunc, "http://google.com/");

  thread1.join();
  thread2.join();

  return 0;
}
```

Is this behaviour a bug?

**Comment by cdglove** [comment]
On Thu, 5 Nov 2015 at 07:41 Andrew <notifications@github.com> wrote:

> I think it is not, because of this line (
> https://github.com/cpp-netlib/uri/blob/143d4264b8db77cf5b77f8975b3ec8c118bb8a80/src/detail/uri_parse.cpp#L250
> ):
>
> bool parse(uri::string_type::iterator first,
>                uri::string_type::iterator last,
>                uri_parts<uri::string_type::iterator> &parts) {
>       namespace qi = boost::spirit::qi;
>       static uri_grammar<uri::string_type> grammar;      <=========== THIS ONE
>       bool is_valid = qi::parse(first, last, grammar, parts);
>       return is_valid && (first == last);
> }
>
> This function is used from uri constructors without any syncronization.
> You can easily reproduce this issue with this code (it is crashing in
> msvc2013 - v120):
>
> void ThreadFunc(const std::string& url) {
>   boost::network::uri::uri t(url);
> }
> int WINAPI _tWinMain(HINSTANCE hInstance, HINSTANCE /*hPrevInstance*/,
>                      LPTSTR lpstrCmdLine,
>                      int nCmdShow) {
>   boost::thread thread1(ThreadFunc, "http://abcd.com/");
>   boost::thread thread2(ThreadFunc, "http://google.com/");
>
>   thread1.join();
>   thread2.join();
>
>   return 0;
> }
>
> Is this behaviour a bug?
>

Does it crash in the grammar constructor or parse? If it crashes in the
grammar constructor it could be said that the issue is actually that MSVC
doesn't correctly synchronize the initialization of statics, which since
c++11 it's supposed to.

-- chris

**Comment by tapapax** [comment]
It crashes while double initializing the same uri_grammar instance. This is because MSVC2013 doesn't support c++11 "magic statics": https://msdn.microsoft.com/en-us/library/hh567368.aspx.

**Comment by tapapax** [comment]
Does it mean that MSVC2013 is not supported by cpp-netlib?

**Comment by deanberris** [comment]
@tapapax No, it means that we might have some work to do here for the 0.11-devel branch. However since we're moving to C++11-only in 0.12-devel, I don't think we can say that we support MSVC 2013 in non-C++11 mode (and since it's a non-standards-compliant compiler, we may never work around it). I'll let @glynos comment on what we're looking to do here, if there's anything that does need fixing.

**Comment by glynos** [comment]
The same problem exists in the master branch (see the uri repo - https://github.com/cpp-netlib/uri/blob/master/src/detail/uri_parse.cpp#L250)

**Comment by deanberris** [comment]
I think the solution here is to not use MSVC 2013 because it's defective with C++11 function-local-statics. If that's not an acceptable solution, I'll accept a pull request to make it work on MSVC 2013 only, maybe through some macro work-arounds (i.e. removing the static in there, and potentially paying the cost on MSVC 2013 builds).

@glynos -- do we know the cost of just removing the static there? I have to admit that when I wrote this a while back I hadn't thought about multi-threading, so potentially just removing the static qualifier would be fine.

**Comment by glynos** [comment]
Removing the static qualifier will make no difference in terms of
behaviour, and will avoid the data race. I don't think constructing the
grammar is an expensive operation, but it does feel pointless to do it
every time on construction. Maybe we could use std::once to initialize the
grammar without adding too much complexity. Dropping MSVC 2013 support
should not be considered as an option.

**Comment by glynos** [comment]
Fixed in commit f590cb3b
